pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
+
+									}
+									//don't want duplicate fields, but want to catch all mapRelationships
+									//over the lineage.
+									if !pairExistsAlready{
+										mapRelationships[mkey] = append(mapRelationships[mkey], pair)
+									}
+								} else {
+									pairs = make([][]string, 0)
+									pairs = append(pairs, pair)
+									mapRelationships[mkey] = pairs
+								}
+
+							}
+						}
+					}
+				}
+			}
 		}
 	}
-	for _, v := range s {
-		if v.AttributeToData[field] == value {
-			return strconv.Itoa(v.Version)
+	return mapRelationships
+}
+//Steps taken in Bisect are:
+//Sort the spec elements in order of their version number.
+
+//Outer loop is going through each of the versions in order.
+//First I parse the query into a slice of field/value pairs.
+//Then build the map of related fields that belong to the same top level attribute
+// and need to be found in the same underlying map. Then I looped over the
+//field value pairs, searched based on the type.
+func (o ObjectLineage) Bisect(argMap map[string]string) string {
+	s := make([]int, 0)
+	for _, value := range o {
+		s = append(s, value.Version)
+	}
+	sort.Ints(s)
+	//get all versions, sort by version, and then build the spec array in order.
+	//if the query conditions are true more than once, want to find the earliest one.
+	specs := make([]Spec, 0)
+	for _, version := range s {
+		specs = append(specs, o[version])
+	}
+	allQueryPairs := make([][]string, 0)
+	//this section is storing the mappings from the query, into the allQueryPairs object
+	//these are the queries, each element must be satisfied somewhere in the spec for the bisect to succeed
+	for key, value := range argMap {
+		attributeValueSlice := make([]string, 2)
+		if strings.Contains(key, "field") {
+			//find associated value in argMap, the query params are such that field1=bar, field2=foo
+			fieldNum, err := strconv.Atoi(key[5:])
+			if err != nil {
+				return fmt.Sprintf("Failure, could not convert %s. Invalid Query parameters.", key)
+			}
+			//find associated value by looking in the map for value+fieldNum.
+			valueOfKey, ok := argMap["value"+strconv.Itoa(fieldNum)]
+			if !ok {
+				return fmt.Sprintf("Could not find an associated value for field: %s", key)
+			}
+			attributeValueSlice[0] = value
+			attributeValueSlice[1] = valueOfKey
+			allQueryPairs = append(allQueryPairs, attributeValueSlice)
 		}
 	}
-	return strconv.Itoa(-1)
+	// fmt.Printf("attributeValuePairs%s\n", allQueryPairs)
+
+	// This method is to build the attributeRelationships from the Query
+	// I Will use this to ensure that fields belonging to the same top-level
+	// attribute, will be treated as a joint query. So you can't just
+	// ask if username ever is daniel and password is ever 223843, because
+	// it could find that in different parts of the spec. They both must be satisfied in the same map object
+	mapRelationships := BuildAttributeRelationships(specs, allQueryPairs)
+	// fmt.Println(mapRelationships)
```

Comment:
Leave this uncommented with following:

fmt.Printf("Query Attributes Map:%v\n", mapRelationships)